### PR TITLE
Improve prefetch behaviour for browsers

### PR DESCRIPTION
**Pull request by bluwy**
## Changes

Closes https://github.com/withastro/astro/issues/10464

- This PR makes all prefetch strategies to use `<link rel="prefetch">` if supported, or else fall back to `fetch()`.
- It also applies the same for the programmatic `prefetch()` API, unless the `with: 'link' | 'fetch'` option is passed. The `with` option is deprecated as it's better to use link whenever possible.

More explanation for why this change is made at https://github.com/withastro/astro/issues/10464#issuecomment-2104238799

## Testing

The existing tests should pass. Testing is a bit tricky because this mainly matters for Firefox and Safari but we don't have those setup. I think manual testing should be enough for now as the prefetch code is isolated and doesn't changed often.

## Docs

https://github.com/withastro/docs/pull/8246

### CODE REVIEW

**Comment by ematipico** [comment]
@bluwy what should we do to test it on Firefox?

**Comment by bluwy** [comment]
We probably need to update `playwright.config.js` `projects` to also test on Firefox etc. But we don't want to run all tests on Firefox, so we may need a second playwright config that only runs specific tests? And I think that should be enough, we can also run it for the view transitions tests.

We also need to update this to install the firefox etc binaries:
https://github.com/withastro/astro/blob/1c5905940241931d93330993f7f62798d19284fd/package.json#L33-L34

**Comment by ematipico** [comment]
Awesome, I can make a PR for that if you're happy

File: packages/astro/src/prefetch/index.ts [hunk 1]
```diff
@@ -215,28 +219,27 @@ export function prefetch(url: string, opts?: PrefetchOptions) {
...
+	// Prefetch with speculationrules if `clientPrerender` is enabled and supported
+	// NOTE: This condition is tree-shaken if `clientPrerender` is false as its a static value
+	if (clientPrerender && HTMLScriptElement.supports?.('speculationrules')) {
+		debug?.(`[astro] Prefetching ${url} with <script type="speculationrules">`);
 		appendSpeculationRules(url);
-	} else if (priority === 'link') {
+	}
+	// Prefetch with link if supported
+	else if (
+		document.createElement('link').relList?.supports?.('prefetch') &&
```
Comment (Princesseuh) [inline packages/astro/src/prefetch/index.ts:230]: Isn't it somewhat expensive to do this on every link? Could it be done once?

Comment (bluwy) [reply packages/astro/src/prefetch/index.ts:230]: I benchmarked this locally and there isn't a big difference in perf from what I can tell compared to caching, so I kept this direct approach, and to save some bytes.

**Review by Princesseuh — APPROVED** [review]

**Review by bholmesdev — APPROVED** [review]

**Review by bholmesdev — APPROVED** [review]
The conditional checks look right to me. Glad we're getting `with` the times!

**Comment by ematipico** [comment]
I added Firefox to the e2e tests, however some of the tests fail e.g. view transitions

**Pull request merged**